File: DelphiHostTemplate/FreeFrame.pas
// www.freeframe.org
// boblists@brightonart.org

{

Copyright (c) 2002, Russell Blakeborough
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

   * Redistributions of source code must retain the above copyright
     notice, this list of conditions and the following disclaimer.
   * Redistributions in binary form must reproduce the above copyright
     notice, this list of conditions and the following disclaimer in
     the documentation and/or other materials provided with the
     distribution.
   * Neither the name of FreeFrame nor the names of its
     contributors may be used to endorse or promote products derived
     from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

See specification.txt in the docs module on FreeFrame CVS at SourcdForge for uptodate function code table and function specifications

}
unit FreeFrame;

interface

uses
  windows, Classes, Controls, SysUtils, dialogs;

type
  TPlugMainFunction = function(functionCode: dword; pParam: pointer; reserved: dword): pointer; stdcall;
  TPluginInfoStruct = record
    APIMajorVersion: dword;
    APIMinorVersion: dword;
    PluginUniqueID: array [0..3] of char;   // 4 characters = 1 dword
    PluginName: array [0..15] of char;       // 16 characters = 4 Dwords
    PluginType: dword; //(effect, source); // woz effect / synth - but could be other than synth eg. live video input
  end;
  TVideoInfoStruct = record
    FrameWidth: dword;
    FrameHeight: dword;
    BitDepth: dword;   // 0=16bit 1=24bit 2=32bit
  end;
  pdw = ^dword;

type
  TFreeFrame = class(TComponent)
  private
    FDirectory: string;
    FPluginIndex: integer;
    FPluginsfound: boolean;
    FPlugins: TStrings;
    FAutoloadplugin: boolean;
    FPluginMajorVersion: integer;
    FPluginMinorVersion: integer;
    FPluginInfo: integer;
    FPluginUniqueID: string;
    FPluginName: string;
    FPluginType: string;
    FSupport16bit : boolean;
    FSupport24bit : boolean;
    FSupport32bit : boolean;
    FNumParameters: integer;
    procedure SetDirectory( Value: string);
    procedure SetPluginIndex( Value: integer);
    procedure SetAutoloadplugin( Value: boolean);
    { Private declarations }
  protected
    { Protected declarations }
  public
    currentPlug : thandle;
    plugMain : tPlugMainFunction;
    PluginInfoStruct: TPluginInfoStruct;
    VideoInfoStruct: TVideoInfoStruct;
    function Getplugins : boolean;
    function Loadplugin : boolean;
    //
    function GetInfo: dword;
    function InitialisePlugin: dword;
    function DeInitialisePlugin: dword;
    function ProcessFrame(pFrame: pointer): dword;
    function GetNumParameters: dword;
    function GetParameterName(Param: dword): string;
    function GetParameterDefault(Param: dword): single;
    function GetParameterDisplay(Param: dword): string;
    function SetParameter(Param: dword; Value: single): dword;
    function GetParameter(Param: dword): single;
    function GetPluginCaps(Param: dword): boolean;
    //
    constructor Create(AOwner: TComponent); override;
    destructor Destroy; override;
  published
    property Pluginsfound : boolean read FPluginsfound;
    property Directory : string read FDirectory write SetDirectory;
    property Plugins : TStrings read FPlugins;
    property PluginIndex : integer read FPluginIndex write SetPluginIndex;
    property Autoloadplugin : boolean read FAutoloadplugin write SetAutoloadplugin default true;
    //
    property PluginMajorVersion : integer read FPluginMajorVersion;
    property PluginInfo : integer read FPluginInfo;
    property PluginMinorVersion : integer read FPluginMinorVersion;
    property PluginUniqueID : string read FPluginUniqueID;
    property PluginName : string read FPluginName;
    property PluginType : string read FPluginType;
    //
    property Support16bit : boolean read FSupport16bit;
    property Support24bit : boolean read FSupport24bit;
    property Support32bit : boolean read FSupport32bit;
    //
    property NumParameters : integer read FNumParameters write FNumParameters;

    { Published declarations }
  end;

procedure Register;

implementation

uses Math;


procedure Register;
begin
  RegisterComponents('FreeFrame', [TFreeFrame]);
end;

{ TFreeFrame }

constructor TFreeFrame.Create(AOwner: TComponent);
begin
  inherited Create(AOwner);
  FPlugins := TStringList.Create;
  currentPlug := 0;
  //
end;

function TFreeFrame.DeInitialisePlugin: dword;
begin
  result:=dword(plugMain(2,nil,0));
end;

destructor TFreeFrame.Destroy;
begin
  FPlugins.Free;
  inherited Destroy;
end;

function TFreeFrame.GetInfo: dword;
var
  dllReturnValue: pointer;
  functionCode: dword;
  pPluginInfoStruct: pointer;
  pParam: PDword;
  tempPChar: PChar;
  i,j: integer;
  tempDword: dword;
  tempString: string;
  tempFCC: array [0..3] of char;
  tempName: array [0..15] of char;
begin
  pPluginInfoStruct:=plugMain(0,nil,0);
  with PluginInfoStruct do begin
    pParam:=pPluginInfoStruct;
    APIMajorVersion:=dword(pParam^);
    inc(pParam);
    APIMinorVersion:=dword(pParam^);
    inc(pParam);
    temppchar:=pchar(pParam);
    for i:=0 to 3 do begin
      PluginUniqueID[i]:=char(tempPchar^);
      inc(temppChar);
    end;
    for i:=0 to 15 do begin
      PluginName[i]:=char(tempPchar^);
      inc(tempPchar);
    end;
    pParam:=pDword(tempPchar);
    PluginType:=dword(pParam^);
  end;
  result:=dword(pPluginInfoStruct);
end;

function TFreeFrame.GetNumParameters: dword;
begin
  result:=dword(plugMain(4,nil,0));
end;

function TFreeFrame.GetParameter(Param: dword): single;
var
  tempDword: dword;
  tempPdword: pdw;
  tempSingle: single;
  tempPsingle: pointer;
begin
  tempdword:=dword(plugMain(9,pointer(Param),0));
  tempPdword:=@tempDword;
  tempPsingle:=@tempSingle;
  copymemory(tempPsingle,tempPdword,4);
  result:=tempSingle;
end;

function TFreeFrame.GetParameterDefault(Param: dword): single;
var
  tempSingle: single;
  tempDword: dword;
begin
  tempDword:=dword(plugMain(6,pointer(Param),0));
  copymemory(@tempSingle,@tempDword,4);
  result:=tempSingle;
end;

function TFreeFrame.GetParameterDisplay(Param: dword): string;
var
  tempParamDisplay: array [0..15] of char;
  tempSourcePointer: pdw;
  tempDestPointer: pdw;
  i: integer;
begin
  tempSourcePointer:=pdw(plugMain(7,pointer(Param),0));
  tempDestPointer:=pdw(@tempParamDisplay);
  for i:=0 to 3 do begin
    tempDestPointer^:=tempSourcePointer^;
    inc(tempSourcePointer);
    inc(tempDestPointer);
  end;
  result:=string(tempParamDisplay);
end;

function TFreeFrame.GetParameterName(Param: dword): string;
var
  tempParamName: array [0..15] of char;
  tempSourcePointer: pdw;
  tempDestPointer: pdw;
  i: integer;
begin
  tempSourcePointer:=pdw(plugMain(5,pointer(Param),0));
  tempDestPointer:=pdw(@tempParamName);
  for i:=0 to 3 do begin
    tempDestPointer^:=tempSourcePointer^;
    inc(tempSourcePointer);
    inc(tempDestPointer);
  end;
  result:=string(tempParamName);
end;

function TFreeFrame.GetPluginCaps(Param: dword): boolean;
begin
  case dword(plugMain(10,pointer(Param),0)) of
    0: result:=false;
    1: result:=true;
  else result:=false;  
  end;
end;

function TFreeFrame.InitialisePlugin: dword;
var
  pVideoInfoStruct: pointer;
begin
  pVideoInfoStruct:=pointer(@VideoInfoStruct);
  result:=dword(plugMain(1,pVideoInfoStruct,0));
end;

function TFreeFrame.ProcessFrame(pFrame: pointer): dword;
begin
  result:=dword(plugMain(3,pFrame,0));
end;

function TFreeFrame.GetPlugins: boolean;
  function checkValid(name: string):boolean;
  const
    functionName = 'plugMain';
  var
    h: thandle;
    proc: pointer;
  begin
    result:=false;
    if not (compareText(copy(name,length(name)-3,4),'.dll')=0) then exit; //findfirst returns filenames containing dll, not just ending
    h := LoadLibrary(pchar(name));
    if h <> 0 then begin //its a dll!
      Proc := GetProcAddress(h, functionName);
      if Proc <> nil then result:=true;
      FreeLibrary(h);
    end;
  end;
var
  t: TSearchRec;
begin
  FPlugins.clear;

  if findfirst(FDirectory+'*.dll', faAnyFile, t) = 0 then begin
    if checkValid(FDirectory+t.name) then FPlugins.add(t.name);
    while findnext(t) = 0 do if checkValid(FDirectory+t.name) then FPlugins.add(t.name);
    findclose(t);
  end;
  if FPlugins.Count>0 then begin
    FPluginsfound := true;
    FPluginIndex := 0;
  end else FPluginsfound := False;

end;


procedure TFreeFrame.SetDirectory( Value: string);
begin
  FDirectory := Value;
end;

function TFreeFrame.SetParameter(Param: dword; Value: single): dword;
type
  TSetParamStruct = array [0..1] of dword;
var
  SetParamStruct: TSetParamStruct;
  tempPdword: pdw;
begin
  SetParamStruct[0]:=Param;
  tempPdword:=@value;
  SetParamStruct[1]:=tempPdword^;
  result:=dword(plugMain(8,@SetParamStruct,0));
end;

procedure TFreeFrame.SetPluginIndex( Value: integer);
begin
  If not FPluginsfound then EXIT;

  //if FPluginIndex = Value then Exit;
  FPluginIndex := Value;

  if FPluginIndex < 0 then exit;

  if currentPlug<>0 then begin
   DeInitialisePlugin;
   freeLibrary(currentPlug);
  end;

  plugMain:=nil;

  currentPlug := LoadLibrary(pchar(FDirectory +'/'+FPlugins.Strings[PluginIndex]));
  if currentPlug <> 0 then begin //its a dll!
    plugMain := GetProcAddress(currentPlug, 'plugMain');
    if @plugMain = nil then begin
      FreeLibrary(currentPlug);
      currentPlug:=0;
    end;
  end;
  if not assigned(plugMain) then showmessage('problem loading plugin');

  if AutoLoadPlugin then begin
    // Load new plugin
    LoadPlugin;
  end;

end;

procedure TFreeFrame.SetAutoloadplugin( Value: boolean);
begin
  FAutoloadplugin := Value;
end;

function TFreeFrame.Loadplugin: boolean;
var
  tempPluginCaps: array [0..2] of boolean;
  i: integer;
begin
  FPluginInfo         := GetInfo;
  FPluginMajorVersion := PluginInfoStruct.APIMajorVersion;
  FPluginMinorVersion := PluginInfoStruct.APIMinorVersion;
  FPluginUniqueID     := PluginInfoStruct.PluginUniqueID;
  FPluginName         := PluginInfoStruct.PluginName;
  case PluginInfoStruct.PluginType of
    0: FPluginType := 'effect';
    1: FPluginType := 'source';
  end;
  // Call GetPluginCaps to see which bitdepths it can manage
  for i:=0 to 2 do tempPluginCaps[i] := GetPluginCaps(i);
  FSupport16bit := tempPluginCaps[0];
  FSupport24bit := tempPluginCaps[1];
  FSupport32bit := tempPluginCaps[2];
  //
  FNumParameters := GetNumParameters;
  InitialisePlugin;
end;

end.

File: MSVCPlugin/FreeFrame.cpp
///////////////////////////////////////////////////////////////////////////////////
// FreeFrame.cpp
//
// FreeFrame Open Video Plugin Prototype
// ANSI C Version

// www.freeframe.org
// marcus@freeframe.org

/*

Copyright (c) 2002, Marcus Clements www.freeframe.org
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

   * Redistributions of source code must retain the above copyright
     notice, this list of conditions and the following disclaimer.
   * Redistributions in binary form must reproduce the above copyright
     notice, this list of conditions and the following disclaimer in
     the documentation and/or other materials provided with the
     distribution.
   * Neither the name of FreeFrame nor the names of its
     contributors may be used to endorse or promote products derived
     from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

*/

///////////////////////////////////////////////////////////////////////////////////////////
//
// includes 
//

#include "FreeFrame.h"
#include <stdio.h>
#include <string.h>

///////////////////////////////////////////////////////////////////////////////////////////
//
// implementation specific definitions
//

#define NUM_PARAMS 3

struct ParameterStruct {
	float value;
	float defaultValue;
	char displayValue[16];
	char name[16];
};

struct VideoPixel24bit {
	BYTE blue;
	BYTE green;
	BYTE red;
};

///////////////////////////////////////////////////////////////////////////////////////////
// Windows DLL Entry point
//
// notes: we may want to capture hModule as the instance of the host...

#ifdef WIN32
BOOL APIENTRY DllMain( HANDLE hModule, 
                       DWORD  ul_reason_for_call, 
                       LPVOID lpReserved
					 )
{
    return TRUE;
}
#endif

///////////////////////////////////////////////////////////////////////////////////////
// plugMain - The one and only exposed function
// parameters: 
//	functionCode - tells the plugin which function is being called
//  pParam - 32-bit parameter or 32-bit pointer to parameter structure
//
// PLUGIN DEVELOPERS:  you shouldn't need to change this function
//
// All parameters are cast as 32-bit untyped pointers and cast to appropriate
// types here
// 
// All return values are cast to 32-bit untyped pointers here before return to 
// the host
//

#ifdef WIN32
 __declspec(dllexport) LPVOID __stdcall plugMain(DWORD functionCode, LPVOID pParam) 
#elif LINUX
   LPVOID plugMain(DWORD functionCode,LPVOID pParam)
#endif	
{
	switch(functionCode) {

	case FF_GETINFO:
		return (LPVOID) getInfo();
	
	case FF_INITIALISE:
		return (LPVOID) initialise( (VideoInfoStruct*) pParam);
	
	case FF_DEINITIALISE:
		return (LPVOID) deInitialise();
	
	case FF_GETNUMPARAMETERS:
		return (LPVOID) getNumParameters();

	case FF_GETPARAMETERNAME:
		return (LPVOID) getParameterName( (DWORD) pParam );
	
	case FF_GETPARAMETERDEFAULT:
		return (LPVOID)(DWORD) getParameterDefault( (DWORD) pParam );
	
	case FF_GETPARAMETERDISPLAY:
		return  (LPVOID) getParameterDisplay( (DWORD) pParam );
	
	// parameters are passed in here as a packed struct of two DWORDS:
	// index and value
	case FF_SETPARAMETER:
		return (LPVOID) setParameter( (SetParameterStruct*) pParam );
	
	case FF_PROCESSFRAME:
		return (LPVOID) processFrame(pParam);

	case FF_GETPARAMETER:
		return (LPVOID)(DWORD) getParameter((DWORD) pParam);

	case FF_GETPLUGINCAPS:
		return (LPVOID) getPluginCaps( (DWORD) pParam);
	
	default:
		return (LPVOID) FF_FAIL;
	}
 }


/////////////////////////////////////////////////////
//
// Plugin Globals
//
// these are defined globally in this sample plugin for simplicities sake
// Plugin developers should allocate memory from the heap for this stuff
//

PlugInfoStruct plugInfo;
VideoInfoStruct videoInfo;
ParameterStruct parameters[NUM_PARAMS];


/////////////////////////////////////////////////////
//
// Sample Function implementation
//

///////////////////////////////////////////////////////////////////////////////////////
// getInfo
//
// gets information about the plugin - version, unique id, short name and type 
// This function should be identical in all future versions of the FreeFrame API
//  
// return values (32-bit pointer to PlugInfoStruct)
// FF_FAIL on error
// 32-bit pointer to PlugInfoStruct
//
// HOST: call this function first to get version information
//       the version defines the other fucntion codes that are supported
//       supported function codes are listed in the documentation www.freeframe.org

PlugInfoStruct* getInfo() 
{
	plugInfo.APIMajorVersion = 0;
	plugInfo.APIMinorVersion = 1024; // lets keep this in sync with the delphi host for now
	char ID[5] = "MPB1";		 // this *must* be unique to your plugin 
								 // see www.freeframe.org for a list of ID's already taken
	char name[17] = "MSVC_Plugin_Base";
	
	memcpy(plugInfo.uniqueID, ID, 4);
	memcpy(plugInfo.pluginName, name, 16);
	plugInfo.pluginType = FF_EFFECT;

	return &plugInfo;
}

///////////////////////////////////////////////////////////////////////////////////////
// initialise
//
// Prepare the Plug-in for processing.  
// Set default values, allocate memory
// When the plug-in returns from this function it must be ready to proces a frame
//
// return values (DWORD)
// FF_SUCCESS - success
// non-zero - fail (error values to be defined)
//
// HOST: This function *must* return before a call to processFrame

DWORD initialise(VideoInfoStruct* pVideoInfo)
{
	// make a copy of the VideoInfoStruct
	videoInfo.frameWidth = pVideoInfo->frameWidth;
	videoInfo.frameHeight = pVideoInfo->frameHeight;
	videoInfo.bitDepth = pVideoInfo->bitDepth;

	// this shouldn't happen if the host is checking the capabilities properly
	if (videoInfo.bitDepth != 1)
	{
		return FF_FAIL;
	}

	// populate the parameters structs
	parameters[0].defaultValue = 0.5f;
	parameters[1].defaultValue = 0.0f;
	parameters[2].defaultValue = 1.0f;
	parameters[0].value = 0.5f;
	parameters[1].value = 0.0f;
	parameters[2].value = 1.0f;
	char tempName1[17] = "brightness";
	char tempName2[17] = "unused____param2";
	char tempName3[17] = "unused____param3";
	memcpy(parameters[0].name, tempName1, 16);
	memcpy(parameters[1].name, tempName2, 16);
	memcpy(parameters[2].name, tempName3, 16);

	return FF_SUCCESS;
}

///////////////////////////////////////////////////////////////////////////////////////
// deinitialise
//
// Tidy up   
// Deallocate memory
//
// return values (DWORD)
// FF_SUCCESS - success
// non-zero - fail (error values to be defined)
//
// HOST: This *must* be the last function called on the plugin

DWORD deInitialise()
{
	return FF_SUCCESS;
}

///////////////////////////////////////////////////////////////////////////////////////
// getNumParameters 
//
// returns number of parameters in plugin
// 
// return values (DWORD)
// number of parameters
// FF_FAIL on error
//

DWORD getNumParameters()
{
	return NUM_PARAMS;  
}

///////////////////////////////////////////////////////////////////////////////////////
// getParameterName
//
// returns pointer to 16 byte char array containing the name of parameter specified by index
//
// parameters:
// DWORD index - index of parameter 
//
// return values (32-bit pointer to char):
// 32-bit pointer to array of char
// FF_FAIL on error
//

char* getParameterName(DWORD index)
{
	return parameters[index].name;
}


///////////////////////////////////////////////////////////////////////////////////////
// getParameterDefault
//
// returns default value of parameter specified by index as 32-bit float 0<=value<=1
//
// parameters:
// DWORD index - index of parameter 
//
// return values (32-bit float):
// 32-bit float value
// FF_FAIL on error
//

float getParameterDefault(DWORD index)
{
	return parameters[index].defaultValue;
}

///////////////////////////////////////////////////////////////////////////////////////
// getParameterDisplay
//
// returns pointer to array of 16 char containing a string to display as the value of
// parameter index
//
// parameters:
// DWORD index - index of parameter 
//
// return values (32-bit pointer to char):
// 32-bit pointer to array of char
// FF_FAIL on error
//

char* getParameterDisplay(DWORD index)
{
	// fill the array with spaces first
	for (int n=0; n<16; n++) {
		parameters[index].displayValue[n] = ' ';
	}
	sprintf(parameters[index].displayValue, "%f",parameters[index].value);
	return parameters[index].displayValue;
}

///////////////////////////////////////////////////////////////////////////////////////
// setParameter
//
// Sets the value of parameter specified by index
// value is a 32-bit float 0<=value<=1
//
// parameters:
// DWORD index - index of parameter 
// 32-bit float value
//
// return values (DWORD):
// FF_FAIL on error
//

DWORD setParameter(SetParameterStruct* pParam)
{
	parameters[pParam->index].value = pParam->value;
	return FF_SUCCESS;
}

///////////////////////////////////////////////////////////////////////////////////////
// getParameter
//
// returns value of parameter specified by index as 32-bit float 0<=value<=1
//
// parameters:
// DWORD index - index of parameter 
//
// return values (32-bit float):
// 32-bit float value
// FF_FAIL on error
//

float getParameter(DWORD index)
{
	return parameters[index].value;
}

///////////////////////////////////////////////////////////////////////////////////////
// processFrame
//
// process a frame of video
//
// parameters:
// 32-bit pointer to byte array containing frame of video
//
// return values (DWORD):
// FF_SUCCESS
// FF_FAIL on error
//

DWORD processFrame(LPVOID pFrame)
{
	VideoPixel24bit* pPixel = (VideoPixel24bit*) pFrame;
	for (DWORD x = 0; x < videoInfo.frameWidth; x++) {
		for (DWORD y = 0; y < videoInfo.frameHeight; y++) {
			pPixel->blue = (BYTE) (pPixel->blue * parameters[0].value);
			pPixel->green = (BYTE) (pPixel->green * parameters[0].value);
			pPixel->red = (BYTE) (pPixel->red * parameters[0].value);
			pPixel++;
		}
	}

	return FF_SUCCESS;
}

///////////////////////////////////////////////////////////////////////////////////////
// getpluginCaps
//
// returns true or false to indicate whether cappable of feature specified by index
//
// parameters:
// DWORD index - index of parameter 
// allowed values:
// 0 - 16 bit video
// 1 - 24 bit video
// 2 - 32 bit video
//
// return values (DWORD):
// FF_TRUE
// FF_FALSE
//

DWORD getPluginCaps(DWORD index)
{
	switch (index) {
	case 0:
		return FF_FALSE;
	case 1:
		return FF_TRUE;
	case 2:
		return FF_FALSE;
	default:
		return FF_FALSE;
	}
}
